stablehlo/dialect/TypeInference.cpp

```diff
@@ -1223,7 +1223,35 @@ static LogicalResult inferGatherReturnTypeComponents(
                             offsetDims, collapsedSliceDims, startIndexMap,
                             indexVectorDim, shape);
 
-  inferredReturnShapes.emplace_back(shape, elementType);
+  // The dimension sizes of result, corresponding to offset dimensions, depend
+  // on attributes (like `collapsed_slice_dims` and `slice_sizes`) and hence are
+  // always static. Whereas, the dimension sizes of result, corresponding to
+  // batch dimensions, depends on input `start_indices` and could be dynamic.
+  // The corresponding bounds, in that case,  are propagated from the
+  // `start_indices`.
+  ArrayRef<int64_t> startIndicesBounds = encodingToBounds(
+      startIndices.getType().cast<RankedTensorType>().getEncoding());
+  SmallVector<int64_t> inferredBounds(resultRank, ShapedType::kDynamic);
+  if (!startIndicesBounds.empty()) {
+    SmallVector<int64_t> batchDims;
+    for (int dim = 0; dim < resultRank; ++dim)
```

Comment:
can we do something like the following? It can avoid find, distance and is_contained.

```
int64_t index = 0;
// make a bit map of isOffsetDim
for (int dim = 0; dim < resultRank; ++dim) {
  if (isOffsetDim(dim)) continue;
  if (index == indexVectorDim) ++index;
  inferredBounds[dim] = startIndicesBounds[index];
  ++index;
}
```

Reply:
Thanks!
@smit-hinsu  PTAL.